### PR TITLE
making HttpUtil class singleton

### PR DESCRIPTION
**Pull request by vityagi**

### CODE REVIEW

File: VMBackup/main/freezesnapshotter.py [hunk 1]
```diff
@@ -114,6 +114,8 @@ def doFreezeSnapshot(self):
         all_failed = False
         unable_to_sleep = False
 
+        http_util = HttpUtil(self.logger)
```
Comment (vityagi) [inline VMBackup/main/freezesnapshotter.py:117]: This is just to ensure HttpUtil singleton object is created before freeze.

File: VMBackup/main/HttpUtil.py [hunk 2]
```diff
@@ -33,23 +33,26 @@
...
-
-        self.tmpFile = './tmp_file_FD76C85E-406F-4CFA-8EB0-CF18B123365C'
+    __instance = None
+    def __new__(cls, hutil):
+        if(cls.__instance is None):
+            hutil.log("Creating HttpUtil")
+            cls.__instance = super(HttpUtil, cls).__new__(cls)
+            try:
+                waagent.MyDistro = waagent.GetMyDistro()
+                Config = waagent.ConfigurationProvider(None)
```
Comment (vityagi) [inline VMBackup/main/HttpUtil.py:43]: On a customer VMs, it seems above waagent APIs were getting stuck during freeze. So by making it singleton, these will not happen during freeze. With this private-fix, backup of customer's 7-8 VMs have started succeeding which were consistently failing for months.

File: VMBackup/main/backuplogger.py [hunk 3]
```diff
@@ -50,7 +50,8 @@ def log(self, msg, local=False, level='Info'):
                 log_msg = self.log_to_con_py3(msg, level)
             else:
                 log_msg = "{0}  {1}  {2} \n".format(str(datetime.datetime.now()) , level , msg)
-                self.log_to_con(log_msg)
+                if(self.enforced_local_flag_value != False):
+                    self.log_to_con(log_msg)
```
Comment (soumyapattnaik) [inline VMBackup/main/backuplogger.py:54]: was this also provided as part of private fix? looks like this could be reason for resolving the issue as well.

Comment (vityagi) [reply VMBackup/main/backuplogger.py:54]: No this was not provided as private fix

File: VMBackup/main/freezesnapshotter.py (hunk 1 above)
Comment (bpramod) [reply VMBackup/main/freezesnapshotter.py:117]: add the same as comment here

File: VMBackup/main/HttpUtil.py [hunk 4]
```diff
@@ -33,23 +33,26 @@
...
-            errorMsg = "Failed to construct ConfigurationProvider, which may due to the old wala code with error: %s, stack trace: %s" % (str(e), traceback.format_exc())
-            hutil.log(errorMsg)
-            Config = None
-        self.logger = hutil
-        if Config != None:
-            self.proxyHost = Config.get("HttpProxy.Host")
-            self.proxyPort = Config.get("HttpProxy.Port")
-
-        self.tmpFile = './tmp_file_FD76C85E-406F-4CFA-8EB0-CF18B123365C'
+    __instance = None
```
Comment (bpramod) [inline VMBackup/main/HttpUtil.py:36]: add a comment here as well - "Singleton class initialization"

File: VMBackup/main/HttpUtil.py [hunk 5]
```diff
@@ -33,23 +33,26 @@
...
+        if(cls.__instance is None):
+            hutil.log("Creating HttpUtil")
+            cls.__instance = super(HttpUtil, cls).__new__(cls)
+            try:
+                waagent.MyDistro = waagent.GetMyDistro()
+                Config = waagent.ConfigurationProvider(None)
+            except Exception as e:
+                errorMsg = "Failed to construct ConfigurationProvider, which may due to the old wala code with error: %s, stack trace: %s" % (str(e), traceback.format_exc())
+                hutil.log(errorMsg)
+                Config = waagent.ConfigurationProvider()
```
Comment (bpramod) [inline VMBackup/main/HttpUtil.py:47]: i remember that we had issue here in initializing config. It was crashing, that is why we removed it from the exception. Is there any reason to add again ?

File: VMBackup/main/HttpUtil.py [hunk 6]
```diff
@@ -33,23 +33,26 @@
...
+            cls.__instance.logger = hutil
+            cls.__instance.proxyHost = Config.get("HttpProxy.Host")
```
Comment (bpramod) [inline VMBackup/main/HttpUtil.py:49]: please have null check on config.
Can you copy the code from the existing code? I see some deviation here.

File: VMBackup/main/freezesnapshotter.py (hunk 1 above)
Comment (vityagi) [reply VMBackup/main/freezesnapshotter.py:118]: Added

File: VMBackup/main/HttpUtil.py (hunk 4 above)
Comment (vityagi) [reply VMBackup/main/HttpUtil.py:36]: Added

File: VMBackup/main/HttpUtil.py (hunk 5 above)
Comment (vityagi) [reply VMBackup/main/HttpUtil.py]: God catch. I think I started doing these changes when earlier changes (removing from exception) were not in released version. I have now made the required changes.

File: VMBackup/main/HttpUtil.py (hunk 6 above)
Comment (vityagi) [reply VMBackup/main/HttpUtil.py]: Done. thanks for pointing.

File: VMBackup/main/HttpUtil.py [hunk 7]
```diff
@@ -33,23 +33,28 @@
...
+            try:
+                waagent.MyDistro = waagent.GetMyDistro()
+                Config = waagent.ConfigurationProvider(None)
+            except Exception as e:
+                errorMsg = "Failed to construct ConfigurationProvider, which may due to the old wala code with error: %s, stack trace: %s" % (str(e), traceback.format_exc())
+                hutil.log(errorMsg)
+                Config = None
+            cls.__instance.logger = hutil
+            if Config != None:
+                cls.__instance.proxyHost = Config.get("HttpProxy.Host")
```
Comment (bpramod) [inline VMBackup/main/HttpUtil.py:51]: you have missed initializing these variables to None.

File: VMBackup/main/HttpUtil.py [hunk 8]
```diff
@@ -33,23 +33,28 @@
...
+                cls.__instance.proxyPort = Config.get("HttpProxy.Port")
+            cls.__instance.tmpFile = './tmp_file_FD76C85E-406F-4CFA-8EB0-CF18B123365C'
+        else:
+            cls.__instance.logger = hutil
```
Comment (bpramod) [inline VMBackup/main/HttpUtil.py:55]: can be removed

Comment (vityagi) [reply VMBackup/main/HttpUtil.py:55]: Kept it intentionally. On customer VM, without this, logger was trying to write to file (enforce flag change was not reflecting).

File: VMBackup/main/HttpUtil.py (hunk 7 above)
Comment (vityagi) [reply VMBackup/main/HttpUtil.py:51]: Too many misses, sorry.

File: VMBackup/main/HttpUtil.py (hunk 8 above)
Comment (bpramod) [reply VMBackup/main/HttpUtil.py:58]: actually logger will still try to write. if we are using hutil.log.
enforce flag is above hutil.log
I am not sure how it is working.

**Pull request merged**